packages/order-watcher/src/order_watcher/order_watcher_websocket.ts

```diff
@@ -0,0 +1,173 @@
+        const broadcastCallback: OnOrderStateChangeCallback = this._broadcastCallback.bind(this);
+        this._orderWatcher.subscribe(broadcastCallback);
+    }
+
+    /**
+     * Activates the WebSocket server by having its HTTP server start listening.
+     */
+    public listen(): void {
+        const port = process.env.ORDER_WATCHER_HTTP_PORT || DEFAULT_HTTP_PORT;
+        this._httpServer.listen(port, () => {
+            logUtils.log(`${new Date()} [Server] Listening on port ${port}`);
+        });
+    }
+
+    /**
+     * Deactivates the WebSocket server by stopping the HTTP server from accepting
+     * new connections.
+     */
+    public close(): void {
+        this._httpServer.close();
+    }
+
+    private async _onMessageCallbackAsync(connection: WebSocket.connection, message: any): Promise<void> {
+        const response: WebSocketResponse = {
+            action: null,
+            success: false,
+            result: null,
+        };
+        try {
+            assert.doesConformToSchema('message', message, webSocketUtf8MessageSchema);
+            const request: WebSocketRequest = JSON.parse(message.utf8Data);
+            assert.doesConformToSchema('request', request, webSocketRequestSchema);
+            response.action = request.action;
+            response.success = true;
+            response.result = await this._routeRequestAsync(request);
+        } catch (err) {
+            response.result = err.toString();
+        }
+        logUtils.log(`${new Date()} [Server] OrderWatcher output: ${JSON.stringify(response)}`);
+        connection.sendUTF(JSON.stringify(response));
+    }
+
+    private _onCloseCallback(connection: WebSocket.connection): void {
+        this._connectionStore.delete(connection);
+        logUtils.log(`${new Date()} [Server] Client ${connection.remoteAddress} disconnected.`);
+    }
+
+    private async _routeRequestAsync(request: WebSocketRequest): Promise<any> {
+        logUtils.log(`${new Date()} [Server] Request received: ${request.action}`);
+        let result = null;
+        switch (request.action) {
+            case OrderWatcherAction.AddOrder: {
+                const signedOrder: SignedOrder = OrderWatcherWebSocketServer._parseSignedOrder(request.signedOrder);
+                await this._orderWatcher.addOrderAsync(signedOrder);
+                break;
+            }
+            case OrderWatcherAction.RemoveOrder: {
+                const orderHash = request.orderHash || '_';
+                this._orderWatcher.removeOrder(orderHash);
+                break;
+            }
+            case OrderWatcherAction.GetStats: {
+                result = this._orderWatcher.getStats();
+                break;
+            }
+            default:
+                // Should never reach here. Should be caught by JSON schema check.
+                throw new Error(`[Server] Invalid request action: ${request.action}`);
```

Comment:
```suggestion
                throw new Error(`[Server] Invalid request action: ${request.action || 'undefined'}`);
```

Reply:
Would this solve the same issue as the `| null` type definition mentioned in your other comment?